src/main/java/teammates/common/util/retry/MaximumRetriesExceededException.java

```diff
@@ -0,0 +1,19 @@
+package teammates.common.util.retry;
+
+@SuppressWarnings("serial")
+public class MaximumRetriesExceededException extends Exception {
+
+    public Object finalResult;
+
+    public MaximumRetriesExceededException(Retryable task) {
+        this(task, null);
+    }
+
+    public MaximumRetriesExceededException(Retryable task, Throwable cause) {
+        super(task.getName() + " failed after maximum retries" + task.finalMessage == null ? "" : ": " + task.finalMessage,
+                cause);
+        if (task instanceof RetryableTaskReturnsThrows) {
+            finalResult = ((RetryableTaskReturnsThrows) task).getResult();
```

Comment:
I decided to simplify it altogether by just copying two optional protected fields from the `Retryable` - `finalMessage` and `finalData`. After all, upon failure the user might want to embed `finalData` that is not necessarily the result.

Ultimately, the result feature should only be used for returning data upon success, and not for anything else (whether it is embedding data in the `MaximumRetriesExceededException` or passing data around from `run` to `isSuccessful` etc. - as you very rightfully pointed out, we can simply declare private variables within the inner class if we want to pass data). With this in mind, the whole framework is simplified greatly.

The latest iteration of `SearchManager` doesn't even need the result feature, and is much cleaner and easier to understand than the previous iteration. Also, structure wise, `putDocumentWithRetry` and `putDocumentsWithRetry` are finally in sync. Thank you for your excellent inputs! :)